File: algorithm/cpp/combinations.cpp
// 77. Combinations
// Difficulty: Medium

// Given two integers n and k, return all possible combinations of k numbers out of 1 ... n.

// For example,
// If n = 4 and k = 2, a solution is:

// [
//   [2,4],
//   [3,4],
//   [2,3],
//   [1,2],
//   [1,3],
//   [1,4],
// ]

// Time : O(n!)
// Space: O(n)

class Solution {
public:
  vector<vector<int> > combine(int n, int k) {
    vector<vector<int> > ans;
    vector<int> path;
    combine(n, k, 1, path, ans);
    return ans;
  }
  void combine(int n, int k, int start, vector<int>& path, vector<vector<int>>& ans) {
    if (k == 0) {
      ans.push_back(path);
      return;
    }

    for(int i = start; i <= n; ++i) {
      path.push_back(i);
      combine(n, k - 1, i + 1, path, ans);
      path.pop_back();
    }
  }
};

File: data-structure-and-algorithm/typescript/leetcode/src/combinations.ts
/*
Given two integers n and k, return all possible combinations of k numbers out of the range [1, n].

You may return the answer in any order.



Example 1:

Input: n = 4, k = 2
Output:
[
  [2,4],
  [3,4],
  [2,3],
  [1,2],
  [1,3],
  [1,4],
]
Example 2:

Input: n = 1, k = 1
Output: [[1]]


Constraints:

1 <= n <= 20
1 <= k <= n
*/

function combine(range: number, count: number): number[][] {
  const result: number[][] = [];
  combineDfs(1, []);
  return result;

  function combineDfs(start: number, path: number[]) {
    if (path.length === count) {
      result.push([...path]);
      return;
    }
    for (let i = start; i <= range; ++i) {
      path.push(i);
      combineDfs(i + 1, path);
      path.pop();
    }
  }
}
